### ISSUE TITLE
Full scan ends but not really (v3.0)

### ISSUE THREAD

**Issue by thiell** [cea-hpc/robinhood#22]
Hi - At the end of a full scan in v3.0, robinhood doesn't tag the scan as finished and displays erroneous last action time.

The log sample below (STATS dumped every 5 min) shows that scan ends on 2016/05/02 at about 12:31:37. At that time, the total number of inodes matches Lustre inodes count so I assume that the scan is indeed finished. Last action suddenly changes from `2016/05/02 12:26:37 (00s ago)` to `2016/04/30 19:36:12 (1.7d ago)`.

```
$ grep STATS /var/log/robinhood/scratch.log | grep 2016/05/02 | egrep 'started at|last action|progress'
<skip>
2016/05/02 12:11:37 [30793/1] STATS |      progress   : 388923887 entries scanned (0 errors)
2016/05/02 12:16:37 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (2.6d ago)
2016/05/02 12:16:37 [30793/1] STATS |      last action: 2016/05/02 12:16:37 (00s ago)
2016/05/02 12:16:37 [30793/1] STATS |      progress   : 389435186 entries scanned (0 errors)
2016/05/02 12:21:37 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (2.6d ago)
2016/05/02 12:21:37 [30793/1] STATS |      last action: 2016/05/02 12:21:37 (00s ago)
2016/05/02 12:21:37 [30793/1] STATS |      progress   : 389945307 entries scanned (0 errors)
2016/05/02 12:26:37 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (2.6d ago)
2016/05/02 12:26:37 [30793/1] STATS |      last action: 2016/05/02 12:26:37 (00s ago)
2016/05/02 12:26:37 [30793/1] STATS |      progress   : 390461583 entries scanned (0 errors)
2016/05/02 12:31:37 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (2.6d ago)
2016/05/02 12:31:37 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (1.7d ago)
2016/05/02 12:31:37 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 12:36:37 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (2.6d ago)
2016/05/02 12:36:37 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (1.7d ago)
2016/05/02 12:36:37 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 12:41:37 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (2.7d ago)
2016/05/02 12:41:37 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (1.7d ago)
2016/05/02 12:41:37 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 12:46:37 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (2.7d ago)
2016/05/02 12:46:37 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (1.7d ago)
2016/05/02 12:46:37 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 12:51:37 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (2.7d ago)
2016/05/02 12:51:37 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (1.7d ago)
2016/05/02 12:51:37 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 12:56:37 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (2.7d ago)
2016/05/02 12:56:37 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (1.7d ago)
2016/05/02 12:56:37 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 13:01:37 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (2.7d ago)
2016/05/02 13:01:37 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (1.7d ago)
2016/05/02 13:01:37 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 13:06:37 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (2.7d ago)
2016/05/02 13:06:37 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (1.7d ago)
2016/05/02 13:06:37 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 13:11:37 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (2.7d ago)
2016/05/02 13:11:37 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (1.7d ago)
2016/05/02 13:11:37 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 13:16:37 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (2.7d ago)
2016/05/02 13:16:37 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (1.7d ago)
2016/05/02 13:16:37 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
<skip>
2016/05/02 22:11:38 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (3.0d ago)
2016/05/02 22:11:38 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (2.1d ago)
2016/05/02 22:11:38 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 22:16:38 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (3.1d ago)
2016/05/02 22:16:38 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (2.1d ago)
2016/05/02 22:16:38 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 22:21:38 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (3.1d ago)
2016/05/02 22:21:38 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (2.1d ago)
2016/05/02 22:21:38 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 22:26:38 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (3.1d ago)
2016/05/02 22:26:38 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (2.1d ago)
2016/05/02 22:26:38 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 22:31:38 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (3.1d ago)
2016/05/02 22:31:38 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (2.1d ago)
2016/05/02 22:31:38 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 22:36:38 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (3.1d ago)
2016/05/02 22:36:38 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (2.1d ago)
2016/05/02 22:36:38 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 22:41:38 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (3.1d ago)
2016/05/02 22:41:38 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (2.1d ago)
2016/05/02 22:41:38 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 22:46:38 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (3.1d ago)
2016/05/02 22:46:38 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (2.1d ago)
2016/05/02 22:46:38 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 22:51:38 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (3.1d ago)
2016/05/02 22:51:38 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (2.1d ago)
2016/05/02 22:51:38 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 22:56:38 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (3.1d ago)
2016/05/02 22:56:38 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (2.1d ago)
2016/05/02 22:56:38 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 23:01:38 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (3.1d ago)
2016/05/02 23:01:38 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (2.1d ago)
2016/05/02 23:01:38 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 23:06:38 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (3.1d ago)
2016/05/02 23:06:38 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (2.1d ago)
2016/05/02 23:06:38 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 23:11:38 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (3.1d ago)
2016/05/02 23:11:38 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (2.1d ago)
2016/05/02 23:11:38 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 23:16:38 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (3.1d ago)
2016/05/02 23:16:38 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (2.2d ago)
2016/05/02 23:16:38 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 23:21:38 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (3.1d ago)
2016/05/02 23:21:38 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (2.2d ago)
2016/05/02 23:21:38 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 23:26:38 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (3.1d ago)
2016/05/02 23:26:38 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (2.2d ago)
2016/05/02 23:26:38 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 23:31:38 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (3.1d ago)
2016/05/02 23:31:38 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (2.2d ago)
2016/05/02 23:31:38 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 23:36:38 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (3.1d ago)
2016/05/02 23:36:38 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (2.2d ago)
2016/05/02 23:36:38 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 23:41:38 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (3.1d ago)
2016/05/02 23:41:38 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (2.2d ago)
2016/05/02 23:41:38 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 23:46:38 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (3.1d ago)
2016/05/02 23:46:38 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (2.2d ago)
2016/05/02 23:46:38 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 23:51:38 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (3.1d ago)
2016/05/02 23:51:38 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (2.2d ago)
2016/05/02 23:51:38 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
2016/05/02 23:56:38 [30793/1] STATS |      started at : 2016/04/29 21:01:25 (3.1d ago)
2016/05/02 23:56:38 [30793/1] STATS |      last action: 2016/04/30 19:36:12 (2.2d ago)
2016/05/02 23:56:38 [30793/1] STATS |      progress   : 390508105 entries scanned (0 errors)
```

As of today, `rbh-report -a` shows:
```
    Last filesystem scan:
            status:          running
            start:           2016/04/29 21:01:25 (9d 13h 38min 52s ago)
            last action:     2016/04/30 19:36:12 (8d 15h 04min 05s ago)

         Statistics:
            entries scanned: 390508105
            errors:          0
            timeouts:        0
            # threads:       8
            average speed:   1860.47 entries/sec
        >>> current speed:   1965.60 entries/sec
```

Version: 3.0 alpha2

To work around this issue, how can I manually flag the scan as done?

**Comment by tl-cea** [comment]
Beside the strange change of "last_action" that needs to be investigated, I assume that looking at pipeline stats you should have a current action running at RM_OLD_ENTRIES step. It is expected to be expensive when you have 0.4 billion entries in your FS (but not as expensive as 8days...). Anyhow a "gstack" could help investigating. If gstack blocks it means rbh is stuck on a kernel call. In this case you can use "crash" (bt) to see what kernel call is hang.

**Comment by thiell** [comment]
RM_OLD_ENTRIES is empty (before and after). I tried with alpha1 and alpha2 and each time, it was after an empty_db (so initial scan). Also I tried to launch robinhood with -F as a workaround but the trigger doesn't start neither. Currently, I have to manually trigger policies using another robinhood instance. gstack doesn't show anything suspect.

**Comment by nzrutman** [comment]
We're also running v3 and find that the policies don't seem to trigger for us either.

**Comment by nzrutman** [comment]
Why does a scan have to complete before running a policy anyhow? IOW, even if it's a partial view of the file system, it should work on what it knows about...

**Comment by tl-cea** [comment]
To NR: this is a security to ensure robinhood is aware of entries that was here before enabling changelog. In particular, this can lead to wrong situations in the case of lhsm_release policy: robinhood will try to release data from a full OST, but it will only know about the new entries - not the old ones, thus leading to release recent data - which we want to avoid.

**Comment by tl-cea** [comment]
I can't reproduce.

My test scenario is:
* empty rbh DB
* create entries in filesystem
* read changelogs :
`robinhood --readlog --once`

So, robinhood is only aware of these new entries and has never scanned the FS.
* run the lhsm_archive policy: `robinhood --run=lhsm_archive -F`
```
2016/05/18 13:59:23 [9574/1] Main | Daemon started (running modules: policy_run(lhsm_archive))
2016/05/18 13:59:23 [9574/2] lhsm_archive | Starting policy run
2016/05/18 13:59:27 [9574/2] lhsm_archive | Policy run summary: time=04s; target=filesystem entries; 276 successful actions (69.00/sec); volume: 0 (0/sec); 199 entries skipped; 0 errors.
``` 

Here is the trigger defined in my config file:
```
lhsm_archive_trigger {
    trigger_on         = scheduled;
    check_interval     = 1min ;
}
```

* Checking the status of entries regarding lhsm_archive (status=new or modified):
```
 # rbh-report --status-info lhsm

lhsm.status    ,     type,      count,   spc_used,     volume,   avg_size
            new,     file,        475,          0,          0,          0
       modified,     file,        133           0,          0,          0
...
```

**Comment by thiell** [comment]
@tl-cea thx! weird!

I have:
```
lhsm_archive_trigger {
    trigger_on = scheduled;                                                                                                                                                                      
    check_interval = 5min;                                                                                                                                                                       
}
```

robinhood is running with:
```
root     28182 19.5  6.4 28355644 25439404 ?   Ssl  May17 205:28 /usr/sbin/robinhood -F --readlog --run=lhsm_archive --detach -d -f /etc/robinhood.d/scratch.conf -p /var/run/rbh.scratch
```
last STATS policy block:
```
2016/05/18 06:53:28 [28182/1] STATS | ======= lhsm_archive policy: trigger stats ======
2016/05/18 06:53:28 [28182/1] STATS | Trigger #0 (scheduled)        : not checked yet.
2016/05/18 06:53:28 [28182/1] STATS |     last run: none
2016/05/18 06:53:28 [28182/1] STATS |     total   : none
2016/05/18 06:53:28 [28182/1] STATS | ======= lhsm_archive policy: action stats ======
2016/05/18 06:53:28 [28182/1] STATS | idle threads       = 4
2016/05/18 06:53:28 [28182/1] STATS | queued actions     = 0
2016/05/18 06:53:28 [28182/1] STATS | action status:
2016/05/18 06:53:28 [28182/1] STATS |     action successful              = 0
2016/05/18 06:53:28 [28182/1] STATS | 0 actions successful/0, 0 (0 blocks)
```

When I run another instance, for example:
```
robinhood --run="lhsm_archive(class:bigfiles,max-count=100K)" --once
```

... then it works. It bet it's something obvious.

**Comment by thiell** [comment]
I also tried:

```
RBH_OPT="--readlog --run=lhsm_archive(class:bigfiles) -F --detach"
```
in /etc/sysconfig/robinhood

In that case the policy is applied but the daemon quits at the end:
```
2016/05/19 08:34:10 [5636/3] lhsm_archive | Warning: could not reach the specified policy target: 587 entries/100000 targeted
2016/05/19 08:34:10 [5636/1] Main | lhsm_archive: policy run terminated (rc = 0).
2016/05/19 08:34:10 [5636/1] Main | All tasks done! Exiting.
```

That means that in my case, both `--run` or `--run=policyname` don't seem to start triggers, but `--run=policyname(class)`  runs the policy immediately but once (not really triggered by lhsm_archive_trigger I guess).

**Comment by tl-cea** [comment]
`Trigger #0 (scheduled)        : not checked yet.`
I cannot reproduce...

Perhaps this will give more info:
`robinhood --run=lhsm_archive -l FULL`

This may also help:
```
lhsm_archive_parameters {
    check_actions_on_startup = no;
}
```

**Comment by thiell** [comment]
Hmm, I can see the following log lines when it doesn't work (not present when doing --run=lhsm_archive(fileclass) ):

```
2016/05/20 08:52:21 [45104/4] Policy | missing attrs to determine if entry [0x200023c53:0x1cff0:0x0] matches scope for policy lhsm_archive: scope_mask=0X4000/0X1/0X4, attr_mask=0X4000/0/0, missing=0/0X1/0X4
2016/05/20 08:52:21 [45104/4] Policy | entry [0x200023c53:0x1cff0:0x0] matches scope for policy alert
```

I'll continue investigating later

**Comment by tl-cea** [comment]
Today's 'master' is nearly beta1. If you can test with it, it will be interesting to see if you still have the issue. Thx

**Comment by thiell** [comment]
I tried with today's master branch, the first thing I noticed was the following warnings are back:

```
2016/05/31 12:27:10 [16395/2] Policy | Attribute is missing for checking fileset 'largedir'
2016/05/31 12:27:10 [16395/2] Policy | Attribute is missing for checking fileset 'hugedir'
```
(a lot of them)

Definition:

```
FileClass largedir {
    definition { type == directory and dircount > 100000 and dircount < 1000000 }
}
FileClass hugedir {
    definition { type == directory and dircount >= 1000000 }
}
```

I managed to get rid of the log messages by adding:
```
    ignore_fileclass = largedir;
    ignore_fileclass = hugedir;
```
to lhsm_archive_rules.

Is that really needed as the scope of lhsm_archive_rules by default is **file**?

The hsm_archive policy doesn't seem to start for now, but it might be too early.

**Comment by tl-cea** [comment]
I understand the source of the warnings: when running a policy, robinhood refreshes fileclasses of entries even for fileclasses that are not referenced in the policy. So, it tries to matches all fileclasses you defined, including largedir/hugedir. The policy scope selects files only, and files don't have a dircount attribute to match the dircount-related fileclasses. That's why you get these warnings.
There is nothing wrong. It should just be more silent in this case. Now I completely undertsand the cause, it should be easy to fix.

Regarding you lhsm_archive trigger issue: could you test with this patch?
https://review.gerrithub.io/#/c/278349

Best regards

**Comment by thiell** [comment]
@tl-cea Great! Merci! ;) I applied the patch and the trigger finally starts (but look at the "Policy run cancelled." early errors). I will let it run this night to see how it goes.

```
2016/06/02 00:01:31 [17947/1] Main | Daemon started (running modules: log_reader,policy_run(lhsm_archive,alert))
2016/06/02 00:01:31 [17947/3] ChangeLog | LU-1331 is fixed in this version of Lustre.
2016/06/02 00:01:31 [17947/2] lhsm_archive | Building policy list - last full FS Scan: 2016/04/29 20:58:54
2016/06/02 00:01:31 [17947/2] lhsm_archive | Starting policy run
2016/06/02 00:01:31 [17947/4] alert | Checking policy rules for filesystem entries
2016/06/02 00:01:31 [17947/4] alert | Building policy list - last full FS Scan: 2016/04/29 20:58:54
2016/06/02 00:01:31 [17947/4] alert | Starting policy run
2016/06/02 00:01:31 [17947/5] EntryProc | CREATE record on already existing entry [0x2000280fd:0x118d7:0x0], path=/scratch/users/<full-path-removed>/flux_2497.txt. This is normal if you scanned it previously.
2016/06/02 00:01:31 [17947/6] EntryProc | CREATE record on already existing entry [0x200026350:0x3530:0x0], path=/scratch/users/<full-path-removed>/0000000002-chr11_pos_bucketed.bam_pos_sorted.bam. This is normal if you scanned it previously.
2016/06/02 00:03:41 [17947/4] alert | Error retrieving list of candidates from database. Policy run cancelled.
2016/06/02 00:03:41 [17947/2] lhsm_archive | Error retrieving list of candidates from database. Policy run cancelled.
2016/06/02 00:03:41 [17947/4] alert | Error running policy on filesystem entries. 0 successful actions; volume: 0; 0 entries skipped; 0 errors.
2016/06/02 00:03:41 [17947/2] lhsm_archive | Error running policy on filesystem entries. 0 successful actions; volume: 0; 0 entries skipped; 0 errors.
2016/06/02 00:04:41 [17947/2] lhsm_archive | Checking policy rules for filesystem entries
2016/06/02 00:04:41 [17947/2] lhsm_archive | Building policy list - last full FS Scan: 2016/04/29 20:58:54
2016/06/02 00:04:41 [17947/2] lhsm_archive | Starting policy run
2016/06/02 00:05:39 [17947/7] ListMgr | Entry has incomplete path in DB: parent_id='0x200027416:0x1ecaf:0x0', relative_path='._Untitled.key'
2016/06/02 00:05:39 [17947/7] EntryProc | Parent dir for entry [0x200027416:0x1ecb1:0x0] is unknown (parent: [0x200027416:0x1ecaf:0x0], child name: '._Untitled.key'): updating entry path info
2016/06/02 00:05:39 [17947/8] ListMgr | Entry has incomplete path in DB: parent_id='0x200027416:0x1ecaf:0x0', relative_path='Untitled.key'
2016/06/02 00:05:39 [17947/8] EntryProc | Parent dir for entry [0x200027416:0x1ecb0:0x0] is unknown (parent: [0x200027416:0x1ecaf:0x0], child name: 'Untitled.key'): updating entry path info
2016/06/02 00:06:31 [17947/1] STATS | ==================== Dumping stats at 2016/06/02 00:06:31 =====================
2016/06/02 00:06:31 [17947/1] STATS | ======== General statistics =========
2016/06/02 00:06:31 [17947/1] STATS | Daemon start time: 2016/06/02 00:01:31
2016/06/02 00:06:31 [17947/1] STATS | Started modules: log_reader,policy_run(lhsm_archive,alert)
2016/06/02 00:06:31 [17947/1] STATS | ChangeLog reader #0:
2016/06/02 00:06:31 [17947/1] STATS |    fs_name    =   regal
2016/06/02 00:06:31 [17947/1] STATS |    mdt_name   =   MDT0000
2016/06/02 00:06:31 [17947/1] STATS |    reader_id  =   cl3
2016/06/02 00:06:31 [17947/1] STATS |    records read        = 20285
2016/06/02 00:06:31 [17947/1] STATS |    interesting records = 20060
2016/06/02 00:06:31 [17947/1] STATS |    suppressed records  = 225
2016/06/02 00:06:31 [17947/1] STATS |    records pending     = 236
2016/06/02 00:06:31 [17947/1] STATS |    last received            = 2016/06/02 00:06:31
2016/06/02 00:06:31 [17947/1] STATS |    last read record time    = 2016/06/02 00:05:59.743810
2016/06/02 00:06:31 [17947/1] STATS |    last read record id      = 9794546983
2016/06/02 00:06:31 [17947/1] STATS |    last pushed record id    = 9794546745
2016/06/02 00:06:31 [17947/1] STATS |    last committed record id = 9794546745
2016/06/02 00:06:31 [17947/1] STATS |    last cleared record id   = 9794546745
2016/06/02 00:06:31 [17947/1] STATS |    read speed               = 10142.50 record/sec (67.62 incl. idle time)
2016/06/02 00:06:31 [17947/1] STATS |    processing speed ratio   = 1.20
2016/06/02 00:06:31 [17947/1] STATS |    ChangeLog stats:
2016/06/02 00:06:31 [17947/1] STATS |    MARK: 0, CREAT: 11020, MKDIR: 690, HLINK: 0, SLINK: 1, MKNOD: 0, UNLNK: 5498, RMDIR: 631
2016/06/02 00:06:31 [17947/1] STATS |    RENME: 214, RNMTO: 0, OPEN: 0, CLOSE: 0, LYOUT: 0, TRUNC: 2081, SATTR: 125, XATTR: 0
2016/06/02 00:06:31 [17947/1] STATS |    HSM: 0, MTIME: 25, CTIME: 0, ATIME: 0
2016/06/02 00:06:31 [17947/1] STATS | ==== EntryProcessor Pipeline Stats ===
2016/06/02 00:06:31 [17947/1] STATS | Idle threads: 32
2016/06/02 00:06:31 [17947/1] STATS | Id constraints count: 0 (hash min=0/max=0/avg=0.0)
2016/06/02 00:06:31 [17947/1] STATS | Name constraints count: 0 (hash min=0/max=0/avg=0.0)
2016/06/02 00:06:31 [17947/1] STATS | Stage              | Wait | Curr | Done |     Total | ms/op |
2016/06/02 00:06:31 [17947/1] STATS |  0: GET_FID        |    0 |    0 |    0 |         0 |  0.00 |
2016/06/02 00:06:31 [17947/1] STATS |  1: GET_INFO_DB    |    0 |    0 |    0 |     20086 |  0.53 |
2016/06/02 00:06:31 [17947/1] STATS |  2: GET_INFO_FS    |    0 |    0 |    0 |     13792 |  2.38 |
2016/06/02 00:06:31 [17947/1] STATS |  3: PRE_APPLY      |    0 |    0 |    0 |     19861 |  0.00 |
2016/06/02 00:06:31 [17947/1] STATS |  4: DB_APPLY       |    0 |    0 |    0 |     19861 |  4.80 |
2016/06/02 00:06:31 [17947/1] STATS |  5: CHGLOG_CLR     |    0 |    0 |    0 |     20086 |  0.02 |
2016/06/02 00:06:31 [17947/1] STATS |  6: RM_OLD_ENTRIES |    0 |    0 |    0 |         0 |  0.00 |
2016/06/02 00:06:31 [17947/1] STATS | DB ops: get=8305/ins=11541/upd=2092/rm=6228
2016/06/02 00:06:31 [17947/1] STATS | ======= lhsm_archive policy: trigger stats ======
2016/06/02 00:06:31 [17947/1] STATS | Trigger #0 (scheduled)        : running.
2016/06/02 00:06:31 [17947/1] STATS |     last run: none
2016/06/02 00:06:31 [17947/1] STATS |     total   : none
2016/06/02 00:06:31 [17947/1] STATS | ======= lhsm_archive policy: action stats ======
2016/06/02 00:06:31 [17947/1] STATS | idle threads       = 4
2016/06/02 00:06:31 [17947/1] STATS | queued actions     = 0
2016/06/02 00:06:31 [17947/1] STATS | action status:
2016/06/02 00:06:31 [17947/1] STATS |     action successful              = 0
2016/06/02 00:06:31 [17947/1] STATS | 0 actions successful/0, 0 (0 blocks)
2016/06/02 00:06:31 [17947/1] STATS | ======= alert policy: trigger stats ======
2016/06/02 00:06:31 [17947/1] STATS | Trigger #0 (scheduled)        : OK (last check: 2016/06/02 00:03:41).
2016/06/02 00:06:31 [17947/1] STATS |     last run: none
2016/06/02 00:06:31 [17947/1] STATS |     total   : none
2016/06/02 00:06:31 [17947/1] STATS | ======= alert policy: action stats ======
2016/06/02 00:06:31 [17947/1] STATS | idle threads       = 4
2016/06/02 00:06:31 [17947/1] STATS | queued actions     = 0
2016/06/02 00:06:31 [17947/1] STATS | action status:
2016/06/02 00:06:31 [17947/1] STATS |     action successful              = 0
2016/06/02 00:06:31 [17947/1] STATS | 0 actions successful/0, 0 (0 blocks)
2016/06/02 00:11:31 [17947/1] STATS | ==================== Dumping stats at 2016/06/02 00:11:31 =====================
2016/06/02 00:11:31 [17947/1] STATS | ======== General statistics =========
2016/06/02 00:11:31 [17947/1] STATS | Daemon start time: 2016/06/02 00:01:31
2016/06/02 00:11:31 [17947/1] STATS | Started modules: log_reader,policy_run(lhsm_archive,alert)
2016/06/02 00:11:31 [17947/1] STATS | ChangeLog reader #0:
2016/06/02 00:11:31 [17947/1] STATS |    fs_name    =   regal
2016/06/02 00:11:31 [17947/1] STATS |    mdt_name   =   MDT0000
2016/06/02 00:11:31 [17947/1] STATS |    reader_id  =   cl3
2016/06/02 00:11:31 [17947/1] STATS |    records read        = 38662
2016/06/02 00:11:31 [17947/1] STATS |    interesting records = 36966
2016/06/02 00:11:31 [17947/1] STATS |    suppressed records  = 1696
2016/06/02 00:11:31 [17947/1] STATS |    records pending     = 221
2016/06/02 00:11:31 [17947/1] STATS |    last received            = 2016/06/02 00:11:31
2016/06/02 00:11:31 [17947/1] STATS |    last read record time    = 2016/06/02 00:11:00.199810
2016/06/02 00:11:31 [17947/1] STATS |    last read record id      = 9794565360
2016/06/02 00:11:31 [17947/1] STATS |    last pushed record id    = 9794565139
2016/06/02 00:11:31 [17947/1] STATS |    last committed record id = 9794565139
2016/06/02 00:11:31 [17947/1] STATS |    last cleared record id   = 9794565139
2016/06/02 00:11:31 [17947/1] STATS |    read speed               = 18377.00 record/sec (61.26 incl. idle time)
2016/06/02 00:11:31 [17947/1] STATS |    processing speed ratio   = 1.00
2016/06/02 00:11:31 [17947/1] STATS |    ChangeLog stats:
2016/06/02 00:11:31 [17947/1] STATS |    MARK: 0, CREAT: 19584, MKDIR: 1804, HLINK: 0, SLINK: 3, MKNOD: 0, UNLNK: 10130, RMDIR: 1144
2016/06/02 00:11:31 [17947/1] STATS |    RENME: 705, RNMTO: 0, OPEN: 0, CLOSE: 0, LYOUT: 0, TRUNC: 4408, SATTR: 858, XATTR: 0
2016/06/02 00:11:31 [17947/1] STATS |    HSM: 0, MTIME: 26, CTIME: 0, ATIME: 0
2016/06/02 00:11:31 [17947/1] STATS | ==== EntryProcessor Pipeline Stats ===
2016/06/02 00:11:31 [17947/1] STATS | Idle threads: 32
2016/06/02 00:11:31 [17947/1] STATS | Id constraints count: 0 (hash min=0/max=0/avg=0.0)
2016/06/02 00:11:31 [17947/1] STATS | Name constraints count: 0 (hash min=0/max=0/avg=0.0)
2016/06/02 00:11:31 [17947/1] STATS | Stage              | Wait | Curr | Done |     Total | ms/op |
2016/06/02 00:11:31 [17947/1] STATS |  0: GET_FID        |    0 |    0 |    0 |         0 |  0.00 |
2016/06/02 00:11:31 [17947/1] STATS |  1: GET_INFO_DB    |    0 |    0 |    0 |     37884 |  0.54 |
2016/06/02 00:11:31 [17947/1] STATS |  2: GET_INFO_FS    |    0 |    0 |    0 |     25570 |  2.45 |
2016/06/02 00:11:31 [17947/1] STATS |  3: PRE_APPLY      |    0 |    0 |    0 |     36461 |  0.00 |
2016/06/02 00:11:31 [17947/1] STATS |  4: DB_APPLY       |    0 |    0 |    0 |     36461 |  4.42 |
2016/06/02 00:11:31 [17947/1] STATS |  5: CHGLOG_CLR     |    0 |    0 |    0 |     37884 |  0.02 |
2016/06/02 00:11:31 [17947/1] STATS |  6: RM_OLD_ENTRIES |    0 |    0 |    0 |         0 |  0.00 |
2016/06/02 00:11:31 [17947/1] STATS | DB ops: get=15347/ins=20809/upd=3828/rm=11824
2016/06/02 00:11:31 [17947/1] STATS | ======= lhsm_archive policy: trigger stats ======
2016/06/02 00:11:31 [17947/1] STATS | Trigger #0 (scheduled)        : running.
2016/06/02 00:11:31 [17947/1] STATS |     last run: none
2016/06/02 00:11:31 [17947/1] STATS |     total   : none
2016/06/02 00:11:31 [17947/1] STATS | ======= lhsm_archive policy: action stats ======                                                                                                           
2016/06/02 00:11:31 [17947/1] STATS | idle threads       = 4                                                                                                                                     
2016/06/02 00:11:31 [17947/1] STATS | queued actions     = 0                                                                                                                                     
2016/06/02 00:11:31 [17947/1] STATS | action status:                                                                                                                                             
2016/06/02 00:11:31 [17947/1] STATS |     action successful              = 0                                                                                                                     
2016/06/02 00:11:31 [17947/1] STATS | 0 actions successful/0, 0 (0 blocks)                                                                                                                       
2016/06/02 00:11:31 [17947/1] STATS | ======= alert policy: trigger stats ======                                                                                                                 
2016/06/02 00:11:31 [17947/1] STATS | Trigger #0 (scheduled)        : OK (last check: 2016/06/02 00:03:41).                                                                                      
2016/06/02 00:11:31 [17947/1] STATS |     last run: none                                                                                                                                         
2016/06/02 00:11:31 [17947/1] STATS |     total   : none                                                                                                                                         
2016/06/02 00:11:31 [17947/1] STATS | ======= alert policy: action stats ======                                                                                                                  
2016/06/02 00:11:31 [17947/1] STATS | idle threads       = 4                                                                                                                                     
2016/06/02 00:11:31 [17947/1] STATS | queued actions     = 0                                                                                                                                     
2016/06/02 00:11:31 [17947/1] STATS | action status:                                                                                                                                             
2016/06/02 00:11:31 [17947/1] STATS |     action successful              = 0                                                                                                                     
2016/06/02 00:11:31 [17947/1] STATS | 0 actions successful/0, 0 (0 blocks)                                                                                                                       
2016/06/02 00:16:18 [17947/9] lhsm_archive | Executing policy action on: 0x20002637f:0x1f3d6:0x0 (/scratch/PI/<full-path-removed>/GRCh37GenomicSuperDup.tab)
<flood of good logs> :)
```

**Comment by tl-cea** [comment]
Good news that your trigger now works!

Your fileclass warnings should be fixed by this one:
    https://review.gerrithub.io/#/c/278624/2

> alert | Error retrieving list of candidates from database. Policy run cancelled.
This suggest a DB connection issue, like a thread or connection starvation => check you DB server log.

**Comment by thiell** [comment]
Yeah, I think you can close this issue when the patch is landed. For me it now works like a charm:

```
2016/06/02 10:06:32 [17947/1] STATS | ======= lhsm_archive policy: action stats ======
2016/06/02 10:06:32 [17947/1] STATS | idle threads       = 4
2016/06/02 10:06:32 [17947/1] STATS | queued actions     = 0
2016/06/02 10:06:32 [17947/1] STATS | action status:
2016/06/02 10:06:32 [17947/1] STATS |     action successful              = 500000
2016/06/02 10:06:32 [17947/1] STATS |     accessed since last update     = 1769
2016/06/02 10:06:32 [17947/1] STATS |     out of scope                   = 115
2016/06/02 10:06:32 [17947/1] STATS | 500000 actions successful/501884, 5.51 TB (11824002120 blocks)
2016/06/02 10:06:32 [17947/1] STATS | last action queued    52 s ago
2016/06/02 10:06:32 [17947/1] STATS | last action started   10 s ago
2016/06/02 10:06:32 [17947/1] STATS | last action completed  9 s ago
```

and with both patches, I don't have any issue left 👍

**Comment by thiell** [comment]
Also you were right re: the DB connection issue. I found out that MySQL's tmpdir got full in some cases when performing on-disk table sorting. I changed tmpdir to a new path with more space.

**Comment by tl-cea** [comment]
Great! It is time to release a beta... nearly a release candidate :)

BTW, we run it now on most of our scratch filesystems. Looks pretty stable.
